ipld/read.go

```diff
@@ -18,190 +17,170 @@ import (
-	// repair the square
-	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(edsWidth) / 2)
-	return rsmt2d.RepairExtendedDataSquare(rowRoots, colRoots, flattened, codec, tree.Constructor)
-}
-
-// uniqueRandNumbers generates count unique random numbers with a max of max
-func uniqueRandNumbers(count, max int) []uint32 {
-	if count > max {
-		panic(fmt.Sprintf("cannot create %d unique samples from a max of %d", count, max))
-	}
-	samples := make(map[uint32]struct{}, count)
-	for i := 0; i < count; {
-		// nolint:gosec // G404: Use of weak random number generator
-		sample := uint32(rand.Intn(max))
-		if _, has := samples[sample]; has {
-			continue
-		}
-		samples[sample] = struct{}{}
-		i++
+	errGroup, ctx := errgroup.WithContext(parentCtx)
+	errGroup.Go(func() error {
+		return fillQuadrant(ctx, quadrant, dag, dataSquare)
+	})
+	if err := errGroup.Wait(); err != nil {
+		return nil, err
 	}
-	out := make([]uint32, count)
-	counter := 0
-	for s := range samples {
-		out[counter] = s
-		counter++
+	batchAdder := NewNmtNodeAdder(parentCtx, ipld.NewBatch(parentCtx, dag, ipld.MaxSizeBatchOption(batchSize(edsWidth))))
+	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(edsWidth)/2, nmt.NodeVisitor(batchAdder.Visit))
+	extended, err := rsmt2d.RepairExtendedDataSquare(dah.RowsRoots, dah.ColumnRoots, dataSquare, codec, tree.Constructor)
+	if err != nil {
+		return nil, err
 	}
-	return out
-}
 
-type index struct {
-	row uint32
-	col uint32
+	return extended, batchAdder.Commit()
 }
 
-type indexedShare struct {
-	data []byte
-	index
+type quadrant struct {
+	isLeftSubtree bool
+	from          int
+	rootCids      []cid.Cid
 }
 
-// shareCounter is a thread safe tallying mechanism for share retrieval
-type shareCounter struct {
-	// all shares
-	shares map[index][]byte
-	// number of shares successfully collected
-	counter uint32
-	// the width of the extended data square
-	edsWidth uint32
-	// the minimum shares needed to repair the extended data square
-	minSharesNeeded uint32
+func pickRandomQuadrant(roots [][]byte) (*quadrant, error) {
+	edsWidth := len(roots)
+	// get the random number from 1 to 4.
+	q := rand.Intn(4) + 1 //nolint:gosec
+	quadrant := &quadrant{rootCids: make([]cid.Cid, edsWidth/2)}
+	// quadrants 1 and 3 corresponds to left subtree,
+	// 2 and 4 to the right subtree
+	/* | 1 | 2 |
+	   | 3 | 4 |
+	*/
+	// choose subtree
+	if q%2 == 1 {
+		quadrant.isLeftSubtree = true
+	}
+	to := 0
+	// define range of shares for sampling
+	if q > 2 {
+		quadrant.from = edsWidth / 2
+		to = edsWidth
+	} else {
+		to = edsWidth / 2
+	}
 
-	shareChan chan indexedShare
-	ctx       context.Context
-	cancel    context.CancelFunc
-	// any errors encountered when attempting to retrieve shares
-	errc chan error
+	var err error
+	for index, counter := quadrant.from, 0; index < to; index++ {
+		quadrant.rootCids[counter], err = plugin.CidFromNamespacedSha256(roots[index])
+		if err != nil {
+			return nil, err
+		}
+		counter++
+	}
```

Comment:
if both `counter` and `index` are incrementing +1, why not just use `index`?

Reply:
because `counter` and `index` could not be equal.

Reply:
Oh sorry, I thought I saw 
```
for index, counter := quadrant.from
```

but it's actually

```
for index, counter := quadrant.from, 0
```